File: lib/types/market.ts
export type MarketDirection = "pump" | "dump";

export type NormalizedCandle = {
  timestamp: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
};

export type NormalizedPricePoint = {
  timestamp: number;
  price: number;
};

export type DataQuality = {
  hasSpotData: boolean;
  hasDerivativesData: boolean;
  hasCandles: boolean;
  hasPriceSeries?: boolean;
  missingFields: string[];
  sourcesUsed: string[];
};

export type NormalizedMarketData = {
  symbol: string;
  name?: string;
  price: number;
  change24h: number;
  volume24h: number;
  marketCap?: number;
  fundingRate?: number;
  openInterest?: number;
  longShortRatio?: number;
  candles?: NormalizedCandle[];
  priceSeries?: NormalizedPricePoint[];
  dataQuality: DataQuality;
};

export type MarketApiError = {
  status: "error";
  message: string;
  details?: string;
};

export type MarketApiSuccess<T> = {
  status: "success";
  data: T;
};

export type MarketApiResponse<T> = MarketApiSuccess<T> | MarketApiError;

export type DecisionName = "LONG_NOW" | "SHORT_NOW" | "SCALP_ONLY" | "WAIT_FOR_CONFIRMATION" | "AVOID";

export type EngineDataQuality = {
  quality: "full" | "partial" | "insufficient";
  notes: string[];
};

export type EngineAnalysis = {
  symbol: string;
  decision: {
    decision: DecisionName;
    confidence: number;
    tradeMode: string;
    riskLevel: "low" | "medium" | "high";
    dominantSide: "bullish" | "bearish" | "neutral";
    invalidationHint: string;
    dominantReasons: string[];
  };
  pressure: {
    buyPressureIntegral: number;
    sellPressureIntegral: number;
    momentumVelocity: number;
    momentumAcceleration: number;
    volatilityPressure: number;
  };
  neuralScores: {
    bullishPressure: number;
    bearishPressure: number;
    trapProbability: number;
    riskScore: number;
    liquidityDominance: number;
    marketStrength: number;
  };
  marketMaking: {
    reservationPrice: number;
    totalSpread: number;
    bidPrice: number;
    askPrice: number;
    liquidityRisk: number;
  };
  scenarios: Array<{
    name: string;
    probability: number;
  }>;
  dataQuality: EngineDataQuality;
};

export type AnalyzeApiSuccess = {
  ok: true;
  source: "python_ai_engine";
  analysis: EngineAnalysis;
};

export type AnalyzeApiError = {
  ok: false;
  reason: "invalid_symbol" | "data_incomplete" | "python_engine_offline" | "python_engine_error" | "market_data_error";
  message: string;
  dataQuality?: unknown;
  details?: string;
};

export type AnalyzeApiResponse = AnalyzeApiSuccess | AnalyzeApiError;

File: ai-engine/app/schemas/market.py
from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class Candle(BaseModel):
    timestamp: int
    open: float
    high: float
    low: float
    close: float
    volume: float


class PricePoint(BaseModel):
    timestamp: int
    price: float


class EngineInputDataQuality(BaseModel):
    quality: Literal["full", "partial", "insufficient"] = "insufficient"
    has_price: bool = Field(default=False, alias="hasPrice")
    has_volume: bool = Field(default=False, alias="hasVolume")
    has_candles: bool = Field(default=False, alias="hasCandles")
    has_price_series: bool = Field(default=False, alias="hasPriceSeries")
    missing: List[str] = Field(default_factory=list)

    model_config = ConfigDict(populate_by_name=True)


class EngineSettings(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    inventory: float = 0.0
    risk_aversion: float = Field(default=0.1, alias="riskAversion")
    time_remaining: float = Field(default=1.0, alias="timeRemaining")
    liquidity_k: float = Field(default=1.5, alias="liquidityK")


class MarketInput(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    symbol: str
    name: Optional[str] = None
    price: float
    change_24h: float = Field(alias="change24h")
    volume_24h: float = Field(alias="volume24h")
    market_cap: Optional[float] = Field(default=None, alias="marketCap")
    funding_rate: Optional[float] = Field(default=None, alias="fundingRate")
    open_interest: Optional[float] = Field(default=None, alias="openInterest")
    long_short_ratio: Optional[float] = Field(default=None, alias="longShortRatio")
    candles: List[Candle] = Field(default_factory=list)
    price_series: List[PricePoint] = Field(default_factory=list, alias="priceSeries")
    data_quality: EngineInputDataQuality = Field(default_factory=EngineInputDataQuality, alias="dataQuality")
    settings: EngineSettings = Field(default_factory=EngineSettings)
